### DISCUSSION TITLE
Lwip ppp - what would I need to do to use ppp instead of wifi as the IP bearer

### DISCUSSION THREAD

**Discussion by johnofleek** [earlephilhower/arduino-pico#1490] [Q&A]
Hi 
I haven't any experience with LWIP
I think that it sits under the arduino-pico wifi functionality?
Will it be a massive task to swap WiFi for ppp?

The reason I ask is that I have a fair amount of experience with cellular / ppp and I'm thinking it might be the easiest way to get a few LTE modules IP connected to the RP2040

Thanks in advance

**Comment by earlephilhower** [comment]
You'd need to implement an lwip_Ethernet-type wrapper, I believe.  Something like was done in https://github.com/esp8266/Arduino/tree/master/libraries/lwIP_PPP (I ~~stole~~borrowed the LWIP<->Ethernet Transport layer infrastructure that @d-a-v built for the ESP8266).

LWIP would also need to be configured to include PPP and `libpico` and the IP stuff rebuilt in `tools/libpico`.  If you're running on a plain Pico (not W), then you may need to introduce a `#define` or something to ensure LWIP is linked in the build.

Overall, I'd say not massive, but not trivial.  It's really limited to getting LWIP and PPP(is it still PPP if it's over serial?) sending packets around.  Once that's done, most of the rest of the WiFi application-style classes should "just work" (but I am sure there are certain gotchas hidden somewhere....).

**Reply by johnofleek** [reply]
Thanks very much for the help that makes a lot of sense. Looking at the esp8266 link I think that should work.

From experience with modems in the past the first gotcha can be avoided if I can get the modems IP WAN up first using AT commands. I should then be able to avoid the bit which breaks most often which is getting the PPP in the host to agree with the PPP in the cellular modem on how to bring up the cellular IP bearer. If I can just use PPP as the IP path between the Pico IP stack and the cellular modems IP stack it has a better chance of working. 

Yes it will be using serial port PPP (PPPos) - hopefully the Pico serial port can be driven reasonably fast?
